modules/javafx.graphics/src/main/java/javafx/scene/Node.java

```diff
@@ -1404,6 +1407,93 @@ public String getName() {
         return visible;
     }
 
+    /**
+     * Indicates whether or not this {@code Node} is shown. A node is considered shown if it's
+     * part of a {@code Scene} that is part of a {@code Window} whose
+     * {@link Window#showingProperty showing property} is {@code true}. The {@link Node#visibleProperty visibility}
+     * of the node or its scene does not affect this property.
+     * <p>
+     * This property can be used in conjunction with {@link ObservableValue#when} to
+     * create bindings which are only actively listening to their source when the node is shown.
+     * <p>
+     * This property can also be used to start animations when the node is shown, and to stop them
+     * when it is no longer shown.
+     *
+     * @since 20
+     */
+    private ReadOnlyBooleanProperty shown;
+
+    public final boolean isShown() {
+        if (shown == null) {  // avoid eager instantiation of property
+            Scene s = getScene();
+            if (s == null) return false;
+            Window w = s.getWindow();
+            return w != null && w.isShowing();
+        }
+
+        return shown.get();
+    }
+
+    public final ReadOnlyBooleanProperty shownProperty() {
+        if (shown == null) {
+            ObservableValue<Boolean> ov = sceneProperty()
+                .flatMap(Scene::windowProperty)
+                .flatMap(Window::showingProperty);  // note: the null case is handled by the delegate with an orElse(false)
+
+            shown = new ReadOnlyBooleanDelegate(Node.this, "shown", ov);
```

Comment:
`shownProperty()` might be used quite a lot, have you considered using a specialized implementation that doesn't need several intermediate observable values?

Reply:
I think that this is already optimized enough by being lazily instantiated (ie. it won't be created if never used).  As a second layer, these are fluent bindings which only register listeners when observed.

Creating a specialized implementation should IMHO replicate the lazy behavior (to avoid too many listeners on window/scene when the property isn't actually observed) which is non-trivial and would require an extensive test. This risks creating bugs now (or later) of which there have been quite a few with similar custom properties -- it's easy to accidentally cache an "old" value or forgetting to clear a cached "current" value when invalidated.

In short, I think optimized custom implementations have been and still are a source of bugs, with no evidence that these optimizations are worth while..